libminifi/src/utils/Id.cpp

```diff
@@ -0,0 +1,154 @@
+  deviceSegment >>= 64 - numBits;
+  logger_->log_debug("Using user defined device segment: %" PRIx64, deviceSegment);
+  deviceSegment <<= 64 - numBits;
+  return deviceSegment;
+}
+
+uint64_t IdGenerator::getRandomDeviceSegment(int numBits) {
+  uint64_t deviceSegment = 0;
+  uuid_t random_uuid;
+  for (int word = 0; word < 2; word++) {
+    uuid_generate_random(random_uuid);
+    for (int i = 0; i < 4; i++) {
+      deviceSegment += random_uuid[i];
+      deviceSegment <<= 8;
+    }
+  }
+  deviceSegment >>= 64 - numBits;
+  logger_->log_debug("Using random device segment: %" PRIx64, deviceSegment);
+  deviceSegment <<= 64 - numBits;
+  return deviceSegment;
+}
+
+void IdGenerator::initialize(const std::shared_ptr<Properties> & properties) {
+  std::string implementation_str;
+  implementation_ = UUID_TIME_IMPL;
+  if (properties->get("uid.implementation", implementation_str)) {
+    std::transform(implementation_str.begin(), implementation_str.end(), implementation_str.begin(), ::tolower);
+    if ("random" == implementation_str) {
+      logger_->log_debug("Using uuid_generate_random for uids.");
+      implementation_ = UUID_RANDOM_IMPL;
+    } else if ("uuid_default" == implementation_str) {
+      logger_->log_debug("Using uuid_generate for uids.");
+      implementation_ = UUID_DEFAULT_IMPL;
+    } else if ("minifi_uid" == implementation_str) {
+      logger_->log_debug("Using minifi uid implementation for uids");
+      implementation_ = MINIFI_UID_IMPL;
+
+      uint64_t timestamp = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch()).count();
+      int device_bits = properties->getInt("uid.minifi.device.segment.bits", 16);
+      std::string device_segment;
+      uint64_t prefix = timestamp;
+      if (device_bits > 0) {
+        if (properties->get("uid.minifi.device.segment", device_segment)) {
+          prefix = getDeviceSegmentFromString(device_segment, device_bits);
+        } else {
+          logger_->log_warn("uid.minifi.device.segment not specified, generating random device segment");
+          prefix = getRandomDeviceSegment(device_bits);
+        }
+        timestamp <<= device_bits;
+        timestamp >>= device_bits;
+        prefix = prefix + timestamp;
+        logger_->log_debug("Using minifi uid prefix: %16" PRIx64, prefix);
+      }
+      for (int i = 0; i < 8; i++) {
+        unsigned char prefix_element = (prefix >> ((7 - i) * 8)) & UNSIGNED_CHAR_MAX;
+        deterministic_prefix_[i] = prefix_element;
+      }
+      incrementor_ = 0;
+    } else if ("time" == implementation_str) {
+      logger_->log_debug("Using uuid_generate_time implementation for uids.");
+    } else {
+      logger_->log_debug("Invalid value for uid.implementation (%s). Using uuid_generate_time implementation for uids.", implementation_str);
+    }
+  } else {
+    logger_->log_debug("Using uuid_generate_time implementation for uids.");
```

Comment:
That documentation question was regarding this log statement and how it seems without the first conditional being true, we simply default to time irrespective of random entry.